### ISSUE TITLE
Engine - KVDB module improvements

### ISSUE THREAD

**Issue by cborla** [wazuh/wazuh#16614]
|Wazuh version|Component|Install type|Install method|Platform|
|---|---|---|---|---|
| 5.0 | Engine | Manager| - | - |

## Description

The primary objective of KVDB is to function as a high-throughput interface for storing key-value pairs permanently.

While the existing implementation satisfies the requirements of the original design, there is room for improvement and the need to address technical debt.

Proposed improvements to the current system include the following:

Improve the DB Handlers to:

1. **Improve the DB Handlers** to:
   1. Count instances.
   2. Identify the Owners.
   3. Automatically unregister the unused handler to facilitate DB closure.
2. **Add resource tuning** at the Manager level, including memory and thread caps.  [Depends on #16702 results]
3. **Allow for raw-level DB operations** such as merging, appending, and replacing entire or partial contents of the DB 
   with raw representations of key-value pairs. Consider implementing patch application.

**Nice to have**:
  1. **Add optional triggers** to subscribe to specific keys and dispatch async handlers when values change. This would allow for triggering specific operations in one module after another module changes relevant information.
  2. ZSH breaks KVDB API when writing a value containing a null character. Might be related to CLI library.

Implementation considerations include:

1. Prioritize the use of `string_view` where possible.
2. Explore the usage of a common transport data type such as an array of bytes.
3. To improve logging in basic or primitive operations, minimize writing to the Wazuh log and return messages to the user for greater control. 
   Include a severity switch to adjust the level of detail in the log output as needed.
   
## Tasks
 - [x] #16702
 - [x] #16704
 - [x] #17056
 - [x] #17128

## DoD
 - [x] Complete all tasks. 
 - [x] Unit test.
 - [x] Update the design document.
 - [x] Update the user manual.
 - [x] Memory Leak analysis.

**Comment by soynof** [comment]
## Decoder to test helper functions

We have created a decoder to be able to test all the helper functions that use the KVDB module.

In order to perform the test, the following must be done:

1. Create an integration and upload the decoder to it:

```
---
name: decoder/test2/0

metadata:
  module: test2
  title: Test logs decoder
  description: Decoder for test
  author:
    name: Wazuh, Inc.
    url: https://wazuh.com
    date: 2023/01/30

sources:
  - decoder/integrations/0

parse:
  logpar:
    # KVDB 30 May 10:04:45 clients connected (0 slaves), 618932 bytes in use, 0 shared objects
    - event.original: <~tmp.id> <@timestamp/%d %b %Y %T>?<@timestamp/%d %b %T><message>

normalize:
  - map:
    # kvdb_set
    - ~tmp.value_1: value_1
    - ~tmp.value_2: value_2
    - ~tmp.json.key_1: value_1
    - ~tmp.json.key_2: value_2
    - ~tmp.set_key_1: +kvdb_set/test_db/key_1/$~tmp.value_1
    - ~tmp.set_key_2: +kvdb_set/test_db/key_2/$~tmp.value_2
    - ~tmp.set_key_3: +kvdb_set/test_db/key_3/$~tmp.json
    - ~tmp.set_key_4: +kvdb_set/test_db/key_4/$~tmp.value_1

    # kvdb_get
    - ~tmp.get_value_1: +kvdb_get/test_db/key_1
    - ~tmp.get_value_2: +kvdb_get/test_db/key_2
    - ~tmp.get_value_3: +kvdb_get/test_db/key_3

    # kvdb_get_merge
    - ~tmp.json_merged.key_3: value_3
    - ~tmp.json_merged.key_4: value_4
    - ~tmp.json_merged: +kvdb_get_merge/test_db/key_3

    - ~tmp.set_host: +kvdb_set/test_db/alfredo/value
    - ~tmp.set_user: +kvdb_set/test_db/pp/value
    - agent.name: alfredo

    # kvdb_delete
    #- ~tmp.delete_key_4: +kvdb_delete/test_db/key_4

  # kvdb_match
  - check:
    - agent.name: +kvdb_match/test_db
    map:
      - ~tmp.host_matched: true

  # kvdb_not_match
  - check:
    - user: +kvdb_not_match/test_db
    map:
      - ~tmp.user_not_matched: true

```

3. Run `/var/ossec/engine/wazuh-engine test`, with the following example event:
 
`KVDB 30 May 10:04:45 clients connected (0 slaves), 618932 bytes in use, 0 shared objects`

5. Validate the returned output, it should be equal to this:

```
OUTPUT:

{
    "wazuh": {
        queue: 49,
        "location": "/dev/stdin",
        "message": "KVDB 30 May 10:04:45 clients connected (0 slaves), 618932 bytes in use, 0 shared objects",
        "decoder": [
            "core-wazuh-message",
            "core-localfile",
            "integrations"
        ],
        "source": "logcollector"
    },
    "event": {
        "original": "KVDB 30 May 10:04:45 clients connected (0 slaves), 618932 bytes in use, 0 shared objects"
    },
    "~tmp": {
        "id": "KVDB",
        "value_1": "value_1",
        "value_2": "value_2",
        "json": {
            "key_1": "value_1",
            "key_2": "value_2"
        },
        "set_key_1": true,
        "set_key_2": true,
        "set_key_3": true,
        "set_key_4": true,
        "get_value_1": "value_1",
        "get_value_2": "value_2",
        "get_value_3": {
            "key_1": "value_1",
            "key_2": "value_2"
        },
        "json_merged": {
            "key_3": "value_3",
            "key_4": "value_4",
            "key_1": "value_1",
            "key_2": "value_2"
        },
        "set_host": true,
        "set_user": true,
        "host_matched": true,
        "user_not_matched": true
    },
    "@timestamp": "2023-05-30T10:04:45.000Z",
    "message": " clients connected (0 slaves), 618932 bytes in use, 0 shared objects",
    agent: {
        "name": "alfredo"
    }
}

```

**Comment by JorgeMarinoDev** [comment]
# KVDB Memory Profiling Report 🟢 


#### Valgrind Execution

##### Command

```bash
valgrind -s --leak-check=full --show-leak-kinds=all --track-origins=yes --log-file=./ValgrindPRs.log.%p --num-callers=15 ./main server --log_level trace --queue_size 1000 --queue_flood_file "" --router_threads 1 start
```

##### Environment

```bash
origin/16614-engine-kvdb-improvements
```

* After rebasing with 13334 including memory leaks fixes on json::Json library and others.
* After upgrading to the latest RocksDB version that included a good amount of mem leak fixes since our last adopted version.

#### Memory Profiling Cases

##### Engine startup gone successfully. No operations on KVDB performed. Just initialization and shutdown.

Three runs with different binary versions where performed.

###### Run 1 : Original Code in branch

```
==56637== LEAK SUMMARY:
==56637==    definitely lost: 0 bytes in 0 blocks
==56637==    indirectly lost: 0 bytes in 0 blocks
==56637==      possibly lost: 0 bytes in 0 blocks
==56637==    still reachable: 12,351 bytes in 158 blocks
==56637==         suppressed: 0 bytes in 0 blocks
```

###### Run 2 : Original Code in branch + Adding DB->Close() call as suggested officially.

```
==58608== LEAK SUMMARY:
==58608==    definitely lost: 0 bytes in 0 blocks
==58608==    indirectly lost: 0 bytes in 0 blocks
==58608==      possibly lost: 0 bytes in 0 blocks
==58608==    still reachable: 12,351 bytes in 158 blocks
==58608==         suppressed: 0 bytes in 0 blocks
```

###### Run 3 : Original Code in branch + Adding DB->Close() and extra call forcing a noop close if anything failed at first run (suggested by a comment in source code, very last resource)

```
==58840== LEAK SUMMARY:
==58840==    definitely lost: 0 bytes in 0 blocks
==58840==    indirectly lost: 0 bytes in 0 blocks
==58840==      possibly lost: 0 bytes in 0 blocks
==58840==    still reachable: 12,351 bytes in 158 blocks
==58840==         suppressed: 0 bytes in 0 blocks
```



All of them returning 

```
still reachable: 12,351 bytes in 158 blocks
```

As the leak category in the report suggests, there are bytes not de-allocated. These are typically associated with a missuse of any Library API at a shutdown stage. Even though several variations on shutdown were tried out obtaining the same old results.

This led to analyzing the 149 blocks in the valgrind's report containing these bytes. 

There are two main categories of blocks in the report:

1. With stack pointing to KVDBManager direcly.
   1. After calliing DBOpen : The DBOptions that RocksDB feeds on, are programatical objects, not static structures. Inside it seems to be some allocated objects, "once", independent of the threads internally used.
2. With stack pointing to RocksDB internals, mostly Thread Implementations in a pool apparently, and some other factories.



#### Conclusion:

There's not much space to do things in another way, being in Initializing or Shutting Down the Rocks DB System. 

Please note, during reasearch, we found a function called **CancelAllBackgroundWork** that's exposed in the API, but only for JAVA. There is no way to call this function explicitly in C/C++. But ! The DB actually calls this method when closing, it means it's already called when shutting down rocksdb the usual way : Deleting the RockDB::DB Object.  (Check DBImpl::~DBImpl() db_impl.cc:753). Several variations already tried. Same results.

##### Engine startup gone successfully. Stress test. Shutdown.

Test

Created 50 databases

```bash
#!/bin/bash

# Loop through the database names from testdb00 to testdb50
for i in {0..50}; do
  # Generate the database name with leading zeros
  database_name=$(printf "testdb%02d" "$i")

  # Execute the command to create the database
  sudo ./main kvdb create -n "$database_name"

  echo "Created database: $database_name"
done
```



Using script to random insert : Via API Inserts random key values of N length and X times in the different 50 dbs.

```bash
#!/bin/bash

# Check if the string length and number of iterations are provided as arguments
if [[ $# -ne 2 ]]; then
  echo "Please provide the string length and number of iterations."
  exit 1
fi

# String length
string_length=$1

# Number of iterations
num_iterations=$2

# Function to generate a random alphanumeric string
generate_random_string() {
  # Length of the random string
  local length=$1

  # Generate a random alphanumeric string using /dev/urandom and base64 encoding
  head /dev/urandom | tr -dc 'a-zA-Z0-9' | head -c "$length"
  echo
}

# Loop through the specified number of iterations
for ((i = 0; i < num_iterations; i++)); do
  # Generate random key and value
  random_key=$(generate_random_string "$string_length")
  random_value=$(generate_random_string "$string_length")

  # Generate random database name with "testdb" as prefix
  database_number=$(printf "%02d" $((RANDOM % 51)))  # Generates a random number between 00 and 50
  database_name="testdb$database_number"

  # Execute the command with the random key, value, and database name
  sudo ./main kvdb insert -n "$database_name" -k "$random_key" -v "$random_value"

  echo "Inserted key: $random_key, value: $random_value into database: $database_name"
done
```



Tests

* 50 insertions of len 12
* 50 insertions of len 20
* 50 insertions of len 50
* 50 insertions of len 200

Results

```
==59335== LEAK SUMMARY:
==59335==    definitely lost: 0 bytes in 0 blocks
==59335==    indirectly lost: 0 bytes in 0 blocks
==59335==      possibly lost: 0 bytes in 0 blocks
==59335==    still reachable: 14,199 bytes in 210 blocks
==59335==         suppressed: 0 bytes in 0 blocks
```

After these operations, a new instance of engine was run. Obtaining same results as at the beginning.

```
==81885== LEAK SUMMARY:
==81885==    definitely lost: 0 bytes in 0 blocks
==81885==    indirectly lost: 0 bytes in 0 blocks
==81885==      possibly lost: 0 bytes in 0 blocks
==81885==    still reachable: 12,351 bytes in 158 blocks
==81885==         suppressed: 0 bytes in 0 blocks
```

Run with only 1 DB and 100 random insertions.

```
==92053== LEAK SUMMARY:
==92053==    definitely lost: 0 bytes in 0 blocks
==92053==    indirectly lost: 0 bytes in 0 blocks
==92053==      possibly lost: 0 bytes in 0 blocks
==92053==    still reachable: 12,375 bytes in 159 blocks
==92053==         suppressed: 0 bytes in 0 blocks
```

Run with only 1 DB and 1000 random insertions.

```
==94183== LEAK SUMMARY:
==94183==    definitely lost: 0 bytes in 0 blocks
==94183==    indirectly lost: 0 bytes in 0 blocks
==94183==      possibly lost: 0 bytes in 0 blocks
==94183==    still reachable: 12,375 bytes in 159 blocks
==94183==         suppressed: 0 bytes in 0 blocks
```



#### Conclusions

Still reachable bytes have a direct proportionallity to the number of KVDBs being open and used. Even more, memcheck reports do not reflect any change based on the number of operations performed on these databases.

These results allow us to reinforce the idea that the still reachable bytes are linked to the data structure itself (number of KVDBs) and not to the operations being run on them over time. 100 insertions throw the same results than 1000 insertions with a single DB. The determinant factor on the number of still reachable bytes is the number of KVDBs.

Going back to the initial results without operations performed on the DBs, these 'bytes' are related to the internal thread pool of rocksDB, which makes full sense if an internal handler must be instantiated for every DB in use. Every thread in charge of a DB has its own 'bytes' still reachable, totally out of the scope of our control at this stage. 

Maybe taking a closer look to the parallelism options that RocksDB offers, we can improve this in case it does not affect how the whole system works.

We can disregard these still reachable bytes at the moment unless a sustained benchmark profiling is done with a period of time of several days or weeks with normal or expected usage of KVDB manager throws any other direct or indirect leak or a heap analysis clearly shows an uprising ramp related to these internal threads of RocksDB. In this case, we can add an intermediate state of our KVDBHandlers reflecting some inactivity window and force these background work to be paused and resumed (Already available in the official RocksDB API).

**Comment by JorgeMarinoDev** [comment]
### KVDB Heap Analysis with massif

#### Run inserting 1000 random key values to 50 random DBS

![image](https://github.com/wazuh/wazuh/assets/117848458/f528c788-d195-4a74-a4fa-6e533646aef6)

We can see overtime, that rocksdb's blocks reach a peak but after on they start decreasing. No ramps detected.


#### Run inserting 500 random key values to 1 DB, creating another DB, inserting 500 random key values and deleting both dbs after all.

### Overall Picture
![image](https://github.com/wazuh/wazuh/assets/117848458/c098f7d0-fda2-4200-a0e3-45a1ce864dd3)


### Creating first DB and inserting

![image](https://github.com/wazuh/wazuh/assets/117848458/ed136d80-d890-4eea-88a5-72135641e4ce)

### Finish inserting in first DB

![image](https://github.com/wazuh/wazuh/assets/117848458/1ae58d04-6c7d-411c-adf9-00271299f604)

### Creating 2nd DB

![image](https://github.com/wazuh/wazuh/assets/117848458/94b7e7b4-b87b-4993-9479-9edcbe37234a)

### Inserting in 2nd DB, deleting both Dbs.

![image](https://github.com/wazuh/wazuh/assets/117848458/944f5408-e303-4826-9987-d40c1f168d08)


#### Conclusions:

There are no observable up-rising ramps related to the amount of operations performed on DBs. Even further, while and after insertions the rocksdb allocators show a decreasing ramp. Related to internal buffers and operation of internal threads only. Memory usage increases when dealing with non already used DBs, but decreases after a regular use even.

Heap over time seems to follow the expected behavior in terms of memory, not related to number of operations or payload size, but strictly related to the number of KVDBs used, even more, after a regular use, the internal buffers accommodate and typically reduce the memory usage after time.

**Comment by cborla** [comment]
Due to the results obtained in the present analysis, we have raised the inquery to rocksdb developers at the following [link](https://github.com/facebook/rocksdb/issues/11598).
